### DISCUSSION TITLE
Frosting: How to stop parallel/background execution of Tasks?

### DISCUSSION THREAD

**Discussion by bergziege** [cake-build/cake#3023] [Q&A] [answered]
Hi and a happy new year to everyone ;-)

I currently having a weird behaviour executing `context.AzurePipelines().Commands.UploadArtifact` which seems to be running async.
How do I know? At least thats how the log tells me:

```
========================================
Upload Artifacts
========================================
Remote Build - Artefakte werden über die Pipeline bereitgestellt.
Upload läuft ...                            <----- Calling context.AzurePipelines().Commands.UploadArtifact here

========================================
UploadArtifactsToTransfer
========================================
An error occurred when executing task 'UploadArtifactsToTransfer'.

----------------------------------------
Teardown
----------------------------------------
Tearing things down...
Error: Error deleting file 'D:\a\1\s\ToBePublished\Some_App-2.4.0.zip'.
System Error.  Code: 32.
The process cannot access the file because it is being used by another process
[...]
Uploading 1 files                         <----- Log output for upload starts here
Total file: 1 ---- Processed file: 0 (0%)
Uploading 'website/Some_App-2.4.0.zip' (5%)
Uploading 'website/Some_App-2.4.0.zip' (10%)
Uploading 'website/Some_App-2.4.0.zip' (15%)
Uploading 'website/Some_App-2.4.0.zip' (20%)
Uploading 'website/Some_App-2.4.0.zip' (25%)
```

My task dependencies are strict sequencial on terms of `Build` -> `Test` -> `Pack` -> `UploadToPipeline` -> `UploadArtifactsToTransfer`.

So how do I wait? for the upload task to finish as `context.AzurePipelines().Commands.UploadArtifact` looks like a blocking call to me?

**Comment by devlead** [comment, marked as answer]
UploadArtifact command only sends a console log command to the Azure DevOps build agent which parses the log and starts a background process that will store and associate the artifact with the build. So it's per definition an asynchronous process.

https://github.com/cake-build/cake/blob/a5cbbfab96b0b8f573d1beeb8664e8d1472759f2/src/Cake.Common/Build/AzurePipelines/AzurePipelinesCommands.cs#L173-L179

https://github.com/cake-build/cake/blob/a5cbbfab96b0b8f573d1beeb8664e8d1472759f2/src/Cake.Common/Build/AzurePipelines/AzurePipelinesCommands.cs#L182-L189

**Reply by pascalberger** [reply]
As @devlead said service messages will always be async. There's a [Cake.AzureDevOps addin](https://cakebuild.net/extensions/cake-azuredevops/) which uses the REST API. The [AzureDevOpsBuild](https://cakebuild.net/api/Cake.AzureDevOps/AzureDevOpsAliases/FBABCD2F) can give you access to the current running build and returns a [AzureDevOpsBuild](https://cakebuild.net/api/Cake.AzureDevOps.Pipelines/AzureDevOpsBuild/) object which then can be used to call to Azure DevOps. I currently only has a [LinkArtifact](https://cakebuild.net/api/Cake.AzureDevOps.Pipelines/AzureDevOpsBuild/1F005595) method, to link to an artifact, but support for publishing artifacts to Azure DevOps could be added in a similar fashion to the addin. I would be happy to accept a PR for this 😄 

Other possibility would be to call the [GetArtifacts](https://cakebuild.net/api/Cake.AzureDevOps.Pipelines/AzureDevOpsBuild/35B3B457) method and wait until the artifact is uploaded, but this would be more of a workaround than a clean solution.
